cli/src/deno.rs

```diff
@@ -1,55 +1,71 @@
 //! Deno runtime for extensions.
 
-use std::fs;
 use std::pin::Pin;
 use std::rc::Rc;
+use std::sync::Arc;
+use std::{fs, thread};
 
 use anyhow::{anyhow, Result};
 use deno_ast::{MediaType, ParseParams, SourceTextInfo};
-use deno_core::{
-    Extension, JsRuntime, ModuleLoader, ModuleSource, ModuleSourceFuture, ModuleSpecifier,
-    ModuleType, RuntimeOptions,
+use deno_runtime::deno_core::{
+    self, ModuleLoader, ModuleSource, ModuleSourceFuture, ModuleSpecifier, ModuleType,
 };
-
-/// Deno runtime state.
-pub struct DenoRuntime {
-    runtime: JsRuntime,
-}
-
-impl Default for DenoRuntime {
-    fn default() -> Self {
-        Self::new()
-    }
-}
-
-impl DenoRuntime {
-    /// Create a new Deno runtime.
-    pub fn new() -> Self {
-        // TODO: Add Phylum API methods here.
-        let phylum_api = Extension::builder().build();
-
-        let runtime = JsRuntime::new(RuntimeOptions {
-            module_loader: Some(Rc::new(TypescriptModuleLoader)),
-            extensions: vec![phylum_api],
-            ..Default::default()
-        });
-
-        Self { runtime }
-    }
-
-    /// Execute a JavaScript module from its main entry point.
-    pub async fn run(&mut self, entrypoint: &str) -> Result<()> {
-        let module_specifier = deno_core::resolve_path(entrypoint)?;
-        let module = self
-            .runtime
-            .load_main_module(&module_specifier, None)
-            .await?;
-        let _ = self.runtime.mod_evaluate(module);
-
-        self.runtime.run_event_loop(false).await?;
-
-        Ok(())
-    }
+use deno_runtime::permissions::Permissions;
+use deno_runtime::worker::{MainWorker, WorkerOptions};
+use deno_runtime::{colors, BootstrapOptions};
+
+/// Execute Phylum extension.
+pub async fn run(entry_point: &str) -> Result<()> {
+    let main_module = deno_core::resolve_path(entry_point)?;
+
+    let cpu_count = thread::available_parallelism()
+        .map(|p| p.get())
+        .unwrap_or(1);
+    let bootstrap = BootstrapOptions {
+        cpu_count,
+        runtime_version: env!("CARGO_PKG_VERSION").into(),
+        user_agent: "phylum-cli/extension".into(),
+        no_color: !colors::use_color(),
+        is_tty: colors::is_tty(),
+        enable_testing_features: Default::default(),
+        debug_flag: Default::default(),
+        ts_version: Default::default(),
+        location: Default::default(),
+        unstable: Default::default(),
+        args: Default::default(),
```

Comment:
Can you not do `..Default::default()` instead?

Reply:
Nope, because the struct doesn't implement `Default`. And `..Default::default()` uses the struct's default implementation, not the default implementation of each field.

Reply:
Gotcha, unfortunate 👍